core/node/rpc/sync/client/syncer_set.go

```diff
@@ -202,57 +202,84 @@ func (ss *SyncerSet) AddStream(
 		return nil // stream is already part of sync operation
 	}
 
-	// check if there is already a syncer that can sync the given stream -> add stream to the syncer
-	if syncer, found := ss.syncers[nodeAddress]; found {
-		if err := syncer.AddStream(ctx, cookie); err != nil {
+	// Returns the target node address and a flag indicating if more nodes are available
+	// Use the node address provided in the sync cookie if available, otherwise use the sticky peer of the stream
+	getNodeAddress := func() (common.Address, *SyncCookie, bool) {
+		return common.BytesToAddress(cookie.GetNodeAddress()), cookie, false
+	}
+	if len(cookie.GetNodeAddress()) == 0 {
+		stream, err := ss.streamCache.GetStreamNoWait(ctx, streamID)
+		if err != nil {
 			return err
 		}
-		ss.streamID2Syncer[streamID] = syncer
-		return nil
-	}
 
-	// first stream to sync with remote -> create a new syncer instance
-	var (
-		syncer StreamsSyncer
-		err    error
-	)
-	if nodeAddress == ss.localNodeAddress {
-		var span trace.Span
-		if ss.otelTracer != nil {
-			_, span = ss.otelTracer.Start(ctx, "NewLocalSyncer",
-				trace.WithAttributes(attribute.String("stream", streamID.String())))
-		}
+		i := 0
+		remotes, useLocal := stream.GetRemotesAndIsLocal()
+		nodeAddress := stream.GetStickyPeer()
+		getNodeAddress = func() (common.Address, *SyncCookie, bool) {
```

Comment:
There is similar functionality in the getStreamImp. Can you extract this functionality into reusable logic?